### PR TITLE
For #11690 - Add areContentsTheSame to top sites and collections in session cont…

### PR DESCRIPTION
**Pull request by ekager**
…rol adapter

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt [hunk 1]
```diff
@@ -38,8 +38,26 @@ import mozilla.components.feature.tab.collections.Tab as ComponentTab
...
-    data class TopSiteList(val topSites: List<TopSite>) : AdapterItem(TopSiteViewHolder.LAYOUT_ID)
+        ButtonTipViewHolder.LAYOUT_ID
+    )
+
+    data class TopSiteList(val topSites: List<TopSite>) : AdapterItem(TopSiteViewHolder.LAYOUT_ID) {
+        override fun getChangePayload(newItem: AdapterItem): Any? {
+            (newItem as TopSiteList).let {
+                // If nothing needs to be updated visually, return a payload
+                data class TopSiteVisual(val title: String)
+                if (newItem.topSites.map { TopSiteVisual(it.title) } == this.topSites.map {
```
Comment (NotWoods) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt:49]:
```kt
val newTopSites = newItem.topSites.asSequence()
val oldTopSites = this.topSites.asSequence()
if (newTopSites.zip(oldTopSites).all { (new, old) -> new.title == old.title })
```

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt [hunk 2]
```diff
@@ -38,8 +38,26 @@ import mozilla.components.feature.tab.collections.Tab as ComponentTab
...
+                        TopSiteVisual(
+                            it.title
+                        )
+                    }) {
+                    return ChangePayload.NO_VISUAL_CHANGE
```
Comment (NotWoods) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt:54]: I think it's more readable to just have if-else here. We should throw if the cast fails.

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt [hunk 3]
```diff
@@ -38,8 +38,26 @@ import mozilla.components.feature.tab.collections.Tab as ComponentTab
 
 sealed class AdapterItem(@LayoutRes val viewType: Int) {
     data class TipItem(val tip: Tip) : AdapterItem(
-        ButtonTipViewHolder.LAYOUT_ID)
-    data class TopSiteList(val topSites: List<TopSite>) : AdapterItem(TopSiteViewHolder.LAYOUT_ID)
+        ButtonTipViewHolder.LAYOUT_ID
+    )
+
+    data class TopSiteList(val topSites: List<TopSite>) : AdapterItem(TopSiteViewHolder.LAYOUT_ID) {
+        override fun getChangePayload(newItem: AdapterItem): Any? {
```
Comment (NotWoods) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt:45]: `getChangePayload` should have a comment mentioning the item type is always the same as the class type

**Comment by ekager** [comment]
Okay sorry I realized that it probably makes more sense to just do this in `areContentsTheSame` 🤔

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt [hunk 4]
```diff
@@ -38,8 +37,17 @@ import mozilla.components.feature.tab.collections.Tab as ComponentTab
...
-        ButtonTipViewHolder.LAYOUT_ID)
-    data class TopSiteList(val topSites: List<TopSite>) : AdapterItem(TopSiteViewHolder.LAYOUT_ID)
+        ButtonTipViewHolder.LAYOUT_ID
+    )
+
+    data class TopSiteList(val topSites: List<TopSite>) : AdapterItem(TopSiteViewHolder.LAYOUT_ID) {
+        override fun contentsSameAs(other: AdapterItem): Boolean {
+            val newTopSites = (other as? TopSiteList)?.topSites?.asSequence() ?: return false
+            val oldTopSites = this.topSites.asSequence()
+            return (newTopSites.zip(oldTopSites).all { (new, old) -> new.title == old.title })
```
Comment (NotWoods) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt:47]:
```suggestion
            return newTopSites.zip(oldTopSites).all { (new, old) -> new.title == old.title }
```

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt [hunk 5]
```diff
@@ -48,32 +56,48 @@ sealed class AdapterItem(@LayoutRes val viewType: Int) {
         val collection: TabCollection,
         val expanded: Boolean
     ) : AdapterItem(CollectionViewHolder.LAYOUT_ID) {
-        override fun sameAs(other: AdapterItem) = other is CollectionItem && collection.id == other.collection.id
+        override fun sameAs(other: AdapterItem) =
+            other is CollectionItem && collection.id == other.collection.id
+
+        override fun contentsSameAs(other: AdapterItem): Boolean {
+            (other as? CollectionItem)?.let {
+                return (it.expanded == this.expanded && it.collection.title == this.collection.title)
```
Comment (NotWoods) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt:64]:
```suggestion
                return it.expanded == this.expanded && it.collection.title == this.collection.title
```

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt (hunk 4 above)
Comment (ekager) [reply app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt]: why have fewer parens when u can have MORE

**Pull request merged**